jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java

```diff
@@ -584,16 +520,22 @@ public boolean handle()
                             }
                             else
                             {
-                                // TODO: do this non-blocking.
-                                // Callback completeCallback = Callback.from(() -> _state.completed(null), _state::completed);
-                                // _state.completing();
-                                try (Blocker.Callback blocker = Blocker.callback())
+
+                                AtomicInteger count = new AtomicInteger();
+                                Callback errorCallback = Callback.from(() ->
                                 {
-                                    // We do not notify ServletRequestListener on this dispatch because it might not
-                                    // be dispatched to an error page, so we delegate this responsibility to the ErrorHandler.
-                                    dispatch(() -> errorHandler.handle(_servletContextRequest, getServletContextResponse(), blocker));
-                                    blocker.block();
-                                }
+                                    if (count.incrementAndGet() == 2)
+                                        _state.scheduleDispatch();
```

Comment:
@lachlan-roberts  I think the count is a bit confusing and it can be done with an AtomicBoolean:
```java
  AtomicBoolean asyncCompletion = new AtomicBoolean(false)
  Callback errorCallback = Callback.from(() ->
  {
      if (!asyncCompletion.compareAndSet(false, true))
        _state.scheduleDispatch() 
  }
  dispatch(() -> errorHandler.handle(_servletContextRequest, getServletContextResponse(), errorCallback));
  if (asyncCompletion.compareAndSet(false, true))                                   
    return false;
```

Reply:
done